File: src/test/java/com/github/johanfredin/bindygenerator/BindyGeneratorTest.java
package com.github.johanfredin.bindygenerator;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.io.File;
import java.io.FileNotFoundException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Map;
import java.util.Scanner;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

@RunWith(JUnit4.class)
public class BindyGeneratorTest {

    private BindyGenerator generator;
    private Path expectedWithColumns;
    private Path expectedWitoutColumns;
    private Path javaSourceFilePath;
    private Path javaSourceFilePath_NoColumns;

    @Before
    public void init() {
        GeneratorConfig config = getConfig();
        File soure = Paths.get("src/test/resources/person.csv").toFile();
        this.expectedWithColumns = Paths.get("src/test/resources/ExpectedResult_WithColumnNames.java");
        this.expectedWitoutColumns = Paths.get("src/test/resources/ExpectedResult_WithoutColumnNames.java");
        this.javaSourceFilePath = Paths.get("src/test/resources/BindyPerson.java");
        this.javaSourceFilePath_NoColumns = Paths.get("src/test/resources/BindyPerson_NoColumns.java");
        this.generator = new BindyGenerator(config, soure, javaSourceFilePath);
    }

    @Test
    public void testGenerate() throws FileNotFoundException {
        this.generator.generate();
        File result = javaSourceFilePath.toFile();
        File expectedFile = expectedWithColumns.toFile();

        String resultContent = "";
        Scanner sc = new Scanner(result);
        while(sc.hasNextLine()) {
            resultContent += sc.nextLine();
        }
        sc.close();

        String expectedFileContent = "";
        sc = new Scanner(expectedFile);
        while(sc.hasNextLine()) {
            expectedFileContent += sc.nextLine();
        }
        sc.close();

        assertEquals("File content is the same", resultContent, expectedFileContent);
    }

    @Test
    public void testGenerate_NoColumnNames() throws FileNotFoundException {
        this.generator.setGeneratorConfig(
                getConfig(";", "Jonsson", FieldMapping.LOWER_CAMEL_CASE,
                        true, true, false, false));

        this.generator.setJavaSourceFilePath(javaSourceFilePath_NoColumns);
        this.generator.generate();
        File result = javaSourceFilePath_NoColumns.toFile();
        File expectedFile = expectedWitoutColumns.toFile();

        String resultContent = "";
        Scanner sc = new Scanner(result);
        while(sc.hasNextLine()) {
            resultContent += sc.nextLine();
        }
        sc.close();

        String expectedFileContent = "";
        sc = new Scanner(expectedFile);
        while(sc.hasNextLine()) {
            expectedFileContent += sc.nextLine();
        }
        sc.close();

        assertEquals("File content is the same", resultContent, expectedFileContent);
    }

    @Test
    public void getFieldMapFromFile_hasHeader_useNumericFieldValues_ObjectTypes() {
        this.generator.setGeneratorConfig(
                getConfig(";", "Jonsson", FieldMapping.LOWER_CAMEL_CASE,
                        true, true, false, true));

        Map<Integer, BindyField> fieldMapFromFile = this.generator.getFieldMapFromFile();
        assertEquals("Size=4", 4, fieldMapFromFile.size());
        assertEquals("Field name=name","name", fieldMapFromFile.get(0).getJavaFieldName());
        assertEquals("Field name=age","age", fieldMapFromFile.get(1).getJavaFieldName());
        assertEquals("Field name=tax","tax", fieldMapFromFile.get(2).getJavaFieldName());
        assertEquals("Field name=mixedBag","mixedBag", fieldMapFromFile.get(3).getJavaFieldName());

        assertEquals("Field type=String",String.class.getSimpleName(), fieldMapFromFile.get(0).getType());
        assertEquals("Field type=Integer",Integer.class.getSimpleName(), fieldMapFromFile.get(1).getType());
        assertEquals("Field type=Float when mixed int and decimal types",
                Float.class.getSimpleName(), fieldMapFromFile.get(2).getType());
        assertEquals("Field type=String when at least one field is a String",
                String.class.getSimpleName(), fieldMapFromFile.get(3).getType());

        assertEquals(0, fieldMapFromFile.get(0).getPos());
        assertEquals(1, fieldMapFromFile.get(1).getPos());
        assertEquals(2, fieldMapFromFile.get(2).getPos());
        assertEquals(3, fieldMapFromFile.get(3).getPos());
    }

    @Test
    public void getFieldMapFromFile_hasHeader_useNumericFieldValues_PrimitiveTypes() {
        this.generator.setGeneratorConfig(
                getConfig(";", "Jonsson", FieldMapping.LOWER_CAMEL_CASE,
                        true, true, true, true));

        Map<Integer, BindyField> fieldMapFromFile = this.generator.getFieldMapFromFile();
        assertEquals("Size=4", 4, fieldMapFromFile.size());
        assertEquals("Field name=name","name", fieldMapFromFile.get(0).getJavaFieldName());
        assertEquals("Field name=age","age", fieldMapFromFile.get(1).getJavaFieldName());
        assertEquals("Field name=tax","tax", fieldMapFromFile.get(2).getJavaFieldName());
        assertEquals("Field name=mixedBag","mixedBag", fieldMapFromFile.get(3).getJavaFieldName());
        assertEquals("Field type=String",String.class.getSimpleName(), fieldMapFromFile.get(0).getType());
        assertEquals("Field type=int","int", fieldMapFromFile.get(1).getType());
        assertEquals("Field type=float when mixed int and decimal types",
                "float", fieldMapFromFile.get(2).getType());
        assertEquals("Field type=String when at least one field is a String",
                String.class.getSimpleName(), fieldMapFromFile.get(3).getType());
        assertEquals(0, fieldMapFromFile.get(0).getPos());
        assertEquals(1, fieldMapFromFile.get(1).getPos());
        assertEquals(2, fieldMapFromFile.get(2).getPos());
        assertEquals(3, fieldMapFromFile.get(3).getPos());
    }

    @Test
    public void testFieldMapFromFile_noHeader_DoNotUseNumericFieldValues() {
        this.generator.setGeneratorConfig(
                getConfig(";", "Jonsson", FieldMapping.AS_IS,
                        false, false, true, true));

        Map<Integer, BindyField> fieldMapFromFile = this.generator.getFieldMapFromFile();
        assertEquals("Size=4", 4, fieldMapFromFile.size());
        assertEquals("Field name=COLUMN_0","COLUMN_0", fieldMapFromFile.get(0).getJavaFieldName());
        assertEquals("Field name=COLUMN_1","COLUMN_1", fieldMapFromFile.get(1).getJavaFieldName());
        assertEquals("Field name=COLUMN_2","COLUMN_2", fieldMapFromFile.get(2).getJavaFieldName());
        assertEquals("Field name=COLUMN_3","COLUMN_3", fieldMapFromFile.get(3).getJavaFieldName());
        assertEquals("Field type=String", String.class.getSimpleName(), fieldMapFromFile.get(0).getType());
        assertEquals("Field type=String", String.class.getSimpleName(), fieldMapFromFile.get(1).getType());
        assertEquals("Field type=String", String.class.getSimpleName(), fieldMapFromFile.get(2).getType());
        assertEquals("Field type=String when at least one field is a String",
                String.class.getSimpleName(), fieldMapFromFile.get(3).getType());
        assertEquals(0, fieldMapFromFile.get(0).getPos());
        assertEquals(1, fieldMapFromFile.get(1).getPos());
        assertEquals(2, fieldMapFromFile.get(2).getPos());
        assertEquals(3, fieldMapFromFile.get(3).getPos());
    }

    @Test
    public void testFieldMapFromFile_noHeader_UseNumericFieldValues_PrimitiveTypes() {
        this.generator.setGeneratorConfig(
                getConfig(";", "Jonsson", FieldMapping.AS_IS,
                        false, true, true, true));

        Map<Integer, BindyField> fieldMapFromFile = this.generator.getFieldMapFromFile();
        assertEquals("Size=4", 4, fieldMapFromFile.size());
        assertEquals("Field name=COLUMN_0","COLUMN_0", fieldMapFromFile.get(0).getJavaFieldName());
        assertEquals("Field name=COLUMN_1","COLUMN_1", fieldMapFromFile.get(1).getJavaFieldName());
        assertEquals("Field name=COLUMN_2","COLUMN_2", fieldMapFromFile.get(2).getJavaFieldName());
        assertEquals("Field name=COLUMN_3","COLUMN_3", fieldMapFromFile.get(3).getJavaFieldName());

        assertEquals("Field type=String",String.class.getSimpleName(), fieldMapFromFile.get(0).getType());
        assertEquals("Field type=int","int", fieldMapFromFile.get(1).getType());
        assertEquals("Field type=float when mixed int and decimal types",
                "float", fieldMapFromFile.get(2).getType());
        assertEquals("Field type=String when at least one field is a String",
                String.class.getSimpleName(), fieldMapFromFile.get(3).getType());

        assertEquals(0, fieldMapFromFile.get(0).getPos());
        assertEquals(1, fieldMapFromFile.get(1).getPos());
        assertEquals(2, fieldMapFromFile.get(2).getPos());
        assertEquals(3, fieldMapFromFile.get(3).getPos());
    }

    @Test
    public void testGetHeaderField_LowerCamelCase() {
        DatasetHeader headerField = this.generator.getHeaderField("APEX_TWIN", FieldMapping.LOWER_CAMEL_CASE);
        assertEquals("should be APEX_TWIN", "APEX_TWIN", headerField.getDataSourceName());
        assertEquals("Should now be apexTwin", "apexTwin", headerField.getJavaFieldName());
    }

    @Test
    public void testGetHeaderField_AsIs_QuoteCharsRemoved() {
        DatasetHeader headerField = this.generator.getHeaderField("'\"MY_MONKEY\"'", FieldMapping.AS_IS);
        assertEquals("Should be MY_MONKEY", "MY_MONKEY", headerField.getDataSourceName());
        assertEquals("Should now be MY_MONKEY", "MY_MONKEY", headerField.getJavaFieldName());
    }

    @Test
    public void testGetHeaderField_LowerCamelCase_IllegalDelimitersRemoved() {
        DatasetHeader headerField = this.generator.getHeaderField("MY-LITTLE MONKEY\"'", FieldMapping.LOWER_CAMEL_CASE);
        assertEquals("Should be MY-LITTLE MONKEY", "MY-LITTLE MONKEY", headerField.getDataSourceName());
        assertEquals("Should now be myLittleMonkey", "myLittleMonkey", headerField.getJavaFieldName());
    }

    @Test(expected = RuntimeException.class)
    public void testGetHeaderField_AsIs_IllegalDelimitersFound() {
        DatasetHeader headerField = this.generator.getHeaderField("MY-LITTLE MONKEY\"'", FieldMapping.AS_IS);
        assertNull("Header field could not be transformed", headerField.getJavaFieldName());
    }

    private GeneratorConfig getConfig() {
        return getConfig(";",
                "BindyPerson",
                FieldMapping.LOWER_CAMEL_CASE,
                true,
                true,
                true,
                true);
    }

    private GeneratorConfig getConfig(String delimiter, String javaClassName, FieldMapping fieldMapping,
                                      boolean isHeader, boolean useNumericFieldTypes,
                                      boolean usePrimitiveTypesWherePossible,
                                      boolean includeColumnName) {
        return new GeneratorConfig(delimiter, javaClassName, fieldMapping, isHeader, useNumericFieldTypes, usePrimitiveTypesWherePossible, includeColumnName);
    }
}

File: src/test/java/com/github/johanfredin/bindygenerator/BindyGeneratorTest.kt
package com.github.johanfredin.bindygenerator

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import java.io.FileNotFoundException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths

@RunWith(JUnit4::class)
class BindyGeneratorTest {

    private val source = Paths.get("src/test/resources/person.csv")
    private val expectedWithColumns = Paths.get("src/test/resources/ExpectedResult_WithColumnNames.java")
    private val expectedWithoutColumns = Paths.get("src/test/resources/ExpectedResult_WithoutColumnNames.java")
    private val javaSourceFilePath = Paths.get("src/test/resources/BindyPerson.java")
    private val javaSourceFilePathNoColumns = Paths.get("src/test/resources/BindyPerson_NoColumns.java")

    @Test
    @Throws(FileNotFoundException::class)
    fun testGenerate() {
        getGenerator().generate()
        verifyResultFileLooksTheSameAsExpected(javaSourceFilePath, expectedWithColumns)
    }

    @Test
    @Throws(FileNotFoundException::class)
    fun testGenerate_NoColumnNames() {
        val config = getConfig(
                delimiter = ";",
                javaClassName = "Jonsson",
                fieldMapping = FieldMapping.LOWER_CAMEL_CASE,
                includeColumnName = false,
                isHeader = true,
                integerType = IntegerType.INTEGER,
                decimalType = DecimalType.FLOAT,
                stringType = `StringType.java`.STRING)

        val generator = getGenerator(config, source, javaSourceFilePathNoColumns)
        generator.generate()

        verifyResultFileLooksTheSameAsExpected(javaSourceFilePathNoColumns, expectedWithoutColumns)
    }

    private fun verifyResultFileLooksTheSameAsExpected(javaSourceFilePath: Path, expectedWithColumns: Path) {

        var resultContent = ""
        Files.lines(javaSourceFilePath)
                .forEach{line -> resultContent += line}
        var expectedFileContent = ""
        Files.lines(expectedWithColumns)
                .forEach { line -> expectedFileContent += line }
        assertEquals("File content is the same", resultContent, expectedFileContent)
    }

    @Test
    fun getFieldMapFromFile_hasHeader_useNumericFieldValues_ObjectTypes() {
        val config = getConfig(
                delimiter = ";",
                javaClassName = "Jonsson",
                fieldMapping = FieldMapping.LOWER_CAMEL_CASE,
                includeColumnName = true,
                isHeader = true,
                integerType = IntegerType.INTEGER,
                decimalType = DecimalType.FLOAT,
                stringType = `StringType.java`.STRING)

        val generator = getGenerator(config)

        val fieldMapFromFile = generator.fieldMapFromFile
        assertEquals("Size=4", 4, fieldMapFromFile.size.toLong())
        assertEquals("Field name=name", "name", fieldMapFromFile[0]?.javaFieldName)
        assertEquals("Field name=age", "age", fieldMapFromFile[1]?.javaFieldName)
        assertEquals("Field name=tax", "tax", fieldMapFromFile[2]?.javaFieldName)
        assertEquals("Field name=mixedBag", "mixedBag", fieldMapFromFile[3]?.javaFieldName)

        assertEquals("Field objectTypeName=String", `StringType.java`.STRING.label, fieldMapFromFile[0]?.type)
        assertEquals("Field objectTypeName=Integer", IntegerType.INTEGER.label, fieldMapFromFile[1]?.type)
        assertEquals("Field objectTypeName=Float when mixed int and decimal types",
                DecimalType.FLOAT.label, fieldMapFromFile[2]?.type)
        assertEquals("Field objectTypeName=String when at least one field is a String",
                `StringType.java`.STRING.label, fieldMapFromFile[3]?.type)

        assertEquals(0, fieldMapFromFile[0]?.pos)
        assertEquals(1, fieldMapFromFile[1]?.pos)
        assertEquals(2, fieldMapFromFile[2]?.pos)
        assertEquals(3, fieldMapFromFile[3]?.pos)
    }

    @Test
    fun getFieldMapFromFile_hasHeader_useNumericFieldValues_PrimitiveTypes() {
        val config = getConfig(
                delimiter = ";",
                javaClassName = "Jonsson",
                fieldMapping = FieldMapping.LOWER_CAMEL_CASE,
                includeColumnName = true,
                isHeader = true,
                integerType = IntegerType.INTEGER_PRIMITIVE,
                decimalType = DecimalType.FLOAT_PRIMITIVE,
                stringType = `StringType.java`.STRING)

        val fieldMapFromFile = getGenerator(config).fieldMapFromFile

        assertEquals("Size=4", 4, fieldMapFromFile.size.toLong())
        assertEquals("Field name=name", "name", fieldMapFromFile[0]?.javaFieldName)
        assertEquals("Field name=age", "age", fieldMapFromFile[1]?.javaFieldName)
        assertEquals("Field name=tax", "tax", fieldMapFromFile[2]?.javaFieldName)
        assertEquals("Field name=mixedBag", "mixedBag", fieldMapFromFile[3]?.javaFieldName)

        assertEquals("Field objectTypeName=String", `StringType.java`.STRING.label, fieldMapFromFile[0]?.type)
        assertEquals("Field objectTypeName=int", IntegerType.INTEGER_PRIMITIVE.label, fieldMapFromFile[1]?.type)
        assertEquals("Field objectTypeName=float when mixed int and decimal types",
                DecimalType.FLOAT_PRIMITIVE.label, fieldMapFromFile[2]?.type)
        assertEquals("Field objectTypeName=String when at least one field is a String",
                `StringType.java`.STRING.label, fieldMapFromFile[3]?.type)

        assertEquals(0, fieldMapFromFile[0]?.pos)
        assertEquals(1, fieldMapFromFile[1]?.pos)
        assertEquals(2, fieldMapFromFile[2]?.pos)
        assertEquals(3, fieldMapFromFile[3]?.pos)
    }

    @Test
    fun testFieldMapFromFile_noHeader_DoNotUseNumericFieldValues() {
        val config = getConfig(
                delimiter = ";",
                javaClassName = "Jonsson",
                fieldMapping = FieldMapping.AS_IS,
                includeColumnName = true,
                isHeader = false,
                integerType = IntegerType.DEFAULT,
                decimalType = DecimalType.DEFAULT,
                stringType = `StringType.java`.STRING)

        val fieldMapFromFile = getGenerator(config).fieldMapFromFile

        assertEquals("Size=4", 4, fieldMapFromFile.size.toLong())
        assertEquals("Field name=COLUMN_0", "COLUMN_0", fieldMapFromFile[0]?.javaFieldName)
        assertEquals("Field name=COLUMN_1", "COLUMN_1", fieldMapFromFile[1]?.javaFieldName)
        assertEquals("Field name=COLUMN_2", "COLUMN_2", fieldMapFromFile[2]?.javaFieldName)
        assertEquals("Field name=COLUMN_3", "COLUMN_3", fieldMapFromFile[3]?.javaFieldName)

        assertEquals("Field objectTypeName=String", `StringType.java`.STRING.label, fieldMapFromFile[0]?.type)
        assertEquals("Field objectTypeName=String", `StringType.java`.STRING.label, fieldMapFromFile[1]?.type)
        assertEquals("Field objectTypeName=String", `StringType.java`.STRING.label, fieldMapFromFile[2]?.type)
        assertEquals("Field objectTypeName=String when at least one field is a String",
                String::class.java.simpleName, fieldMapFromFile[3]?.type)

        assertEquals(0, fieldMapFromFile[0]?.pos)
        assertEquals(1, fieldMapFromFile[1]?.pos)
        assertEquals(2, fieldMapFromFile[2]?.pos)
        assertEquals(3, fieldMapFromFile[3]?.pos)
    }

    @Test
    fun testFieldMapFromFile_noHeader_UseNumericFieldValues_PrimitiveTypes() {
        val config = getConfig(
                delimiter = ";",
                javaClassName = "Jonsson",
                fieldMapping = FieldMapping.AS_IS,
                includeColumnName = true,
                isHeader = false,
                integerType = IntegerType.INTEGER_PRIMITIVE,
                decimalType = DecimalType.FLOAT_PRIMITIVE,
                stringType = `StringType.java`.STRING)

        val fieldMapFromFile = getGenerator(config).fieldMapFromFile

        assertEquals("Size=4", 4, fieldMapFromFile.size.toLong())
        assertEquals("Field name=COLUMN_0", "COLUMN_0", fieldMapFromFile[0]?.javaFieldName)
        assertEquals("Field name=COLUMN_1", "COLUMN_1", fieldMapFromFile[1]?.javaFieldName)
        assertEquals("Field name=COLUMN_2", "COLUMN_2", fieldMapFromFile[2]?.javaFieldName)
        assertEquals("Field name=COLUMN_3", "COLUMN_3", fieldMapFromFile[3]?.javaFieldName)

        assertEquals("Field objectTypeName=String", `StringType.java`.STRING.label, fieldMapFromFile[0]?.type)
        assertEquals("Field objectTypeName=int", IntegerType.INTEGER_PRIMITIVE.label, fieldMapFromFile[1]?.type)
        assertEquals("Field objectTypeName=float when mixed int and decimal types",
                DecimalType.FLOAT_PRIMITIVE.label, fieldMapFromFile[2]?.type)
        assertEquals("Field objectTypeName=String when at least one field is a String",
                `StringType.java`.STRING.label, fieldMapFromFile[3]?.type)

        assertEquals(0, fieldMapFromFile[0]?.pos)
        assertEquals(1, fieldMapFromFile[1]?.pos)
        assertEquals(2, fieldMapFromFile[2]?.pos)
        assertEquals(3, fieldMapFromFile[3]?.pos)
    }

    @Test
    fun testGetHeaderField_LowerCamelCase() {
        val headerField = getGenerator().getHeaderField("APEX_TWIN", FieldMapping.LOWER_CAMEL_CASE)
        assertEquals("should be APEX_TWIN", "APEX_TWIN", headerField.dataSourceName)
        assertEquals("Should now be apexTwin", "apexTwin", headerField.javaFieldName)
    }

    @Test
    fun testGetHeaderField_AsIs_QuoteCharsRemoved() {
        val headerField = getGenerator().getHeaderField("'\"MY_MONKEY\"'", FieldMapping.AS_IS)
        assertEquals("Should be MY_MONKEY", "MY_MONKEY", headerField.dataSourceName)
        assertEquals("Should now be MY_MONKEY", "MY_MONKEY", headerField.javaFieldName)
    }

    @Test
    fun testGetHeaderField_LowerCamelCase_IllegalDelimitersRemoved() {
        val headerField = getGenerator().getHeaderField("MY-LITTLE MONKEY\"'", FieldMapping.LOWER_CAMEL_CASE)
        assertEquals("Should be MY-LITTLE MONKEY", "MY-LITTLE MONKEY", headerField.dataSourceName)
        assertEquals("Should now be myLittleMonkey", "myLittleMonkey", headerField.javaFieldName)
    }

    @Test(expected = RuntimeException::class)
    fun testGetHeaderField_AsIs_IllegalDelimitersFound() {
        val headerField = getGenerator().getHeaderField("MY-LITTLE MONKEY\"'", FieldMapping.AS_IS)
        assertNull("Header field could not be transformed", headerField.javaFieldName)
    }

    @Test
    fun testGetFieldType_PrimitiveTypes() {
        val generator = getGenerator()
        assertEquals(FieldType.STRING, generator.getType("Hello"))
        assertEquals(FieldType.INTEGER, generator.getType("25"))
        assertEquals(FieldType.DECIMAL, generator.getType("24.5"))
    }

    @Test
    fun testGetFieldType_ObjectTypes() {
        val config = getConfig(
                delimiter = ";",
                javaClassName = "Jonsson",
                fieldMapping = FieldMapping.AS_IS,
                includeColumnName = true,
                isHeader = true,
                integerType = IntegerType.INTEGER,
                decimalType = DecimalType.FLOAT,
                stringType = `StringType.java`.STRING)
        val generator = getGenerator(config)
        assertEquals(FieldType.STRING, generator.getType("Hello"))
        assertEquals(FieldType.INTEGER, generator.getType("25"))
        assertEquals(FieldType.DECIMAL, generator.getType("24.5"))
    }

    @Test
    fun testMappingToEnumCorrect() {
        val config = getConfig(
                integerType = IntegerType.INTEGER,
                decimalType = DecimalType.FLOAT,
                stringType = `StringType.java`.STRING)
        val generator = getGenerator(config)
        assertEquals(`StringType.java`.STRING.label, generator.mapToCorrespondingFieldType(generator.getType("Hello")))
        assertEquals(IntegerType.INTEGER.label, generator.mapToCorrespondingFieldType(generator.getType("25")))
        assertEquals(DecimalType.FLOAT.label, generator.mapToCorrespondingFieldType(generator.getType("24.5")))
    }

    @Test
    fun testMappingToEnumCorrect_primitiveTypes() {
        val config = getConfig(
                integerType = IntegerType.SHORT_PRIMITIVE,
                decimalType = DecimalType.DOUBLE_PRIMITIVE,
                stringType = `StringType.java`.CHAR_SEQUENCE)
        val generator = getGenerator(config)
        assertEquals(`StringType.java`.CHAR_SEQUENCE.label, generator.mapToCorrespondingFieldType(generator.getType("Hello")))
        assertEquals(IntegerType.SHORT_PRIMITIVE.label, generator.mapToCorrespondingFieldType(generator.getType("25")))
        assertEquals(DecimalType.DOUBLE_PRIMITIVE.label, generator.mapToCorrespondingFieldType(generator.getType("24.5")))
    }

    private fun getConfig(delimiter: String = ";",
                          javaClassName: String = "BindyPerson",
                          fieldMapping: FieldMapping = FieldMapping.LOWER_CAMEL_CASE,
                          isHeader: Boolean = true,
                          includeColumnName: Boolean = true,
                          decimalType: DecimalType = DecimalType.FLOAT_PRIMITIVE,
                          integerType: IntegerType = IntegerType.INTEGER_PRIMITIVE,
                          stringType: `StringType.java` = `StringType.java`.STRING): GeneratorConfig {
        return GeneratorConfig(delimiter, javaClassName, fieldMapping, isHeader, includeColumnName, integerType, decimalType, stringType)
    }

    private fun getConfig(
            decimalType: DecimalType,
            integerType: IntegerType,
            stringType: `StringType.java`): GeneratorConfig {
        return GeneratorConfig(delimiter = ";",
                javaClassName = "Jonsson",
                fieldMapping = FieldMapping.AS_IS,
                isIncludeColumnName = true,
                isHeader = true,
                decimalType = decimalType,
                integerType = integerType,
                stringType = stringType)
    }

    private fun getGenerator(generatorConfig: GeneratorConfig = getConfig(),
                             source: Path = this.source,
                             javaFilePath: Path = this.javaSourceFilePath): BindyGenerator {
        return BindyGenerator(generatorConfig, source, javaFilePath)
    }
}
